File: algorithms/cpp/longestIncreasingPathinaMatrix/longestIncreasingPathinaMatrix.cpp
// Source: https://leetcode.com/problems/longest-increasing-path-in-a-matrix/
// Author: Miao Zhang
// Date:   2021-02-02

class Solution {
public:
    int longestIncreasingPath(vector<vector<int>>& matrix) {
        if (matrix.empty()) return 0;
        int res = 0;
        int m = matrix.size();
        int n = matrix[0].size();
        vector<vector<int>> dp(m, vector<int>(n, -1));
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                int path = dfs(matrix, dp, i, j);
                res = max(res, path);
            }
        }
        return res;
    }
    
    int dfs(vector<vector<int>> &matrix, vector<vector<int>> dp, int i, int j) {
        if (dp[i][j] != -1) return dp[i][j];
        vector<pair<int, int>> dirs{{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        dp[i][j] = 1;
        for (auto d: dirs) {
            int x = i + d.first;
            int y = j + d.second;
            if (x < 0 || x >= matrix.size() || y < 0 || y >= matrix[0].size() || matrix[x][y] <= matrix[i][j]) continue;
            dp[i][j] = max(dp[i][j], 1 + dfs(matrix, dp, x, y));
        }
        return dp[i][j];
    }
};

File: algorithms/python/longestIncreasingPathinaMatrix/longestIncreasingPathinaMatrix.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/longest-increasing-path-in-a-matrix/
# Author: Miao Zhang
# Date:   2021-02-02

class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        if not matrix or not matrix[0]: return 0
        m = len(matrix)
        n = len(matrix[0])
        res = 0
        dp = [[-1 for _ in range(n)] for _ in range(m)]
        for i in range(m):
            for j in range(n):
                path = self.dfs(matrix, i, j, dp)
                res = max(res, path)
        return res
    
    def dfs(self, matrix: List[List[int]], i: int, j: int, dp: List[List[int]]) -> int:
        if dp[i][j] != -1:
            return dp[i][j]
        res = 1
        dirs = [(1, 0), (-1, 0), (0, 1), (0, -1)]
        for d in dirs:
            x = i + d[0]
            y = j + d[1]
            if x < 0 or x >= len(matrix) or y < 0 or y >= len(matrix[0]) or matrix[x][y] <= matrix[i][j]:
                continue
            path = 1 + self.dfs(matrix, x, y, dp)
            res = max(res, path)
        dp[i][j] = res
        return dp[i][j]
